juju/juju-controller
# Write controller config and signal the controller agent

This patch builds on the handler added to the controller configuration socket in https://github.com/juju/juju/pull/17002.

It supersedes #67, which in turn was based on #60.

This patch does two things:
- We no longer assume that the identity of the controller itself is congruent with the controller unit ID. We now get this information from the running controller jujud process.
- When the controller configuration file is changed, we send a reload request to the controller's configuration socket, causing a reload of the new configuration.

Note that this assumes that the charm _does not_ deploy the controller itself, which is how we ultimately want to manage HA control planes.

To verify:
- `charmcraft pack` this charm.
- Bootstrap a new LXD controller using `--controller-charm-path <path-to-packed-artefact>`.
- `juju enable-ha`.
- Once all units have joined the relation, ssh to the controller machine and check _/var/lib/juju/agents/controller-0/agent.conf_. It should be populated along the lines of:
```
db-bind-addresses:
  controller/0: 10.246.27.181
  controller/1: 10.246.27.53
  controller/2: 10.246.27.12
```
- `juju debug-log -m controller --replay|grep reload` will indicate signals across the different controllers.

https://warthogs.atlassian.net/browse/JUJU-5306
